### PR TITLE
Feature/emit livewire event after validation failure

### PR DESCRIPTION
**Pull request by chrisrhymes**
1️⃣ Is this something that is wanted/needed? Did you create an issue / discussion about it first?
I created a discussion, but there wasn't any responses, so apologies if this is too much of a niche request, but I thought it would be useful to others to easily display a toast message each time validation fails. 

2️⃣ Does it contain multiple, unrelated changes? Please separate the PRs out.
Single change

3️⃣ Does it include tests, if possible? (Not a deal-breaker, just a nice-to-have)
A simple test has been added, but could probably be improved

4️⃣ Please include a thorough description of the improvement and reasons why it's useful.
You can listen for the failed-validation [event in JavaScript](https://laravel-livewire.com/docs/2.x/events#in-js) to display a toast message. This may not be needed on short forms, but on larger forms that require scrolling it will allow a message to be displayed to users each time they press submit on the form to inform them of validation errors. 

5️⃣ Thanks for contributing! 🙌

### CODE REVIEW

**Comment by iksaku** [comment]
Wouldn't it be better to surround the event emit in a `try... catch`, emit then re-throw the exception?

Just tinkering to prevent double validation execution, since you evaluate for a `fail` in the if statement, and, independent of the result, you then re-validate.

Couldn't this be a considerable hit for complex validation rules or bigger ones?

What do you think?

**Comment by chrisrhymes** [comment]
Thanks for your feedback @iksaku I have done as you have suggested and wrapped in a try catch.

File: src/ComponentConcerns/ValidatesInput.php [hunk 1]
```diff
@@ -153,7 +153,12 @@ public function validate($rules = null, $messages = [], $attributes = [])
 
         $this->shortenModelAttributes($data, $rules, $validator);
 
-        $validatedData = $validator->validate();
+        try {
+            $validatedData = $validator->validate();
+        } catch (ValidationException $e) {
+            $this->emit('failed-validation');
+            throw new ValidationException($validator);
```
Comment (iksaku) [inline src/ComponentConcerns/ValidatesInput.php:160]: To reduce memory usage, you can simply rethrow the catched exception: `throw $e`.

**Review by iksaku — CHANGES_REQUESTED** [review]

File: src/ComponentConcerns/ValidatesInput.php (hunk 1 above)
Comment (chrisrhymes) [reply src/ComponentConcerns/ValidatesInput.php]: Thanks for your help and patience. Every day is a school day.

**Comment by crnkovic** [comment]
I'd avoid this in core of Livewire, couples too much to your application code. I'd rather leave you do it yourself in your components.

**Comment by chrisrhymes** [comment]
Now I understand I can do this in my component without putting it in the Livewire core. Thanks for the help @iksaku and @crnkovic

**Pull request closed without merging**